### PR TITLE
Refactoring Login, Nodes, Metrics and Overview Pages with Setup Script and Type Script

### PR DESCRIPTION
**Pull request by Kinplemelon**

### CODE REVIEW

File: src/views/Dashboard/Nodes.vue [hunk 1]
```diff
@@ -103,50 +103,51 @@
...
+
+const { t } = useI18n();
 
-    onMounted(() => {
-      allNodes();
-      allStats();
-    });
+let nodes: Ref<Array<NodeMsg>> = ref([]);
+let stats: Ref<Array<NodeStatisticalData>> = ref([]);
+let nodesLockTable = ref(true);
```
Comment (kerry-emqx) [inline src/views/Dashboard/Nodes.vue:126]: 简单类型的待遇不如引用类型？为何前面有，后面无

File: src/views/Dashboard/Nodes.vue [hunk 2]
```diff
@@ -103,50 +103,51 @@
...
+const allNodes = async () => {
+  nodes.value = (await loadNodes().catch(() => {})) ?? [];
+  nodesLockTable.value = false;
+};
+const allStats = async () => {
+  stats.value = (await loadStats().catch(() => {})) ?? [];
+  statsLockTable.value = false;
+};
+const caseInsensitiveCompare = (
+  w: undefined | string,
```
Comment (kerry-emqx) [inline src/views/Dashboard/Nodes.vue:141]: why undefined

File: src/views/Dashboard/components/NodesGraph.vue [hunk 3]
```diff
@@ -61,194 +61,196 @@
...
-    let checkNode = (event, n) => {
-      currentInfo.value = n;
-    };
+let nodes: Ref<Array<NodeMsg>> = ref([]);
+let stats: Ref<Array<NodeStatisticalData>> = ref([]);
+let nodeStat: Ref<Record<string, NodeStatisticalData>> = ref({});
+let graph: Ref<undefined | HTMLElement> = ref(undefined);
+let currentInfo: Ref<Array<Record<string, number | string>>> = ref([{}, {}]);
+let infoLoading = ref(true);
+let svg: Ref<any> = ref({});
```
Comment (kerry-emqx) [inline src/views/Dashboard/components/NodesGraph.vue:91]: any?

File: src/views/Dashboard/components/NodesGraph.vue [hunk 4]
```diff
@@ -61,194 +61,196 @@
...
 
-    const tl = function (key, collection = "Dashboard") {
-      return this.$t(collection + "." + key);
-    };
+let getNodes = async () => {
+  let res: Array<NodeMsg> | void = await loadNodes().catch(() => {});
```
Comment (kerry-emqx) [inline src/views/Dashboard/components/NodesGraph.vue:94]: 同前面一样的问题

File: src/views/Dashboard/components/SimpleLine.vue [hunk 5]
```diff
@@ -12,121 +22,113 @@ import "echarts/lib/component/title";
...
+  },
+});
 
-  props: {
-    value: {
-      type: Object,
-      default: () => ({ x: [], y: [] }),
-      required: false,
+const chartEl = ref();
+const chart: Ref<any> = ref(null);
```
Comment (kerry-emqx) [inline src/views/Dashboard/components/SimpleLine.vue:44]: any？

**Review by kerry-emqx — APPROVED** [review]

File: src/api/common.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  NodeMsg,
+  NodeStatisticalData,
+  Prometheus,
+  StatsD,
+} from "@/types/dashboard";
+
+//account
+export function login(
+  user: {
+    password?: string;
```
[the remark this answers is not in the document]
Comment (Kinplemelon) [reply src/api/common.ts:14]: 👌🏻

File: src/views/Base/Login.vue [hunk 7]
```diff
@@ -6,7 +6,7 @@
       loginKeepWidth && 'login-align-width',
     ]"
   >
-    <el-card shadow="never" class="login-card emq-list-card" id="login">
+    <el-card shadow="never" class="login-card emq-list-card" ref="loginCom">
```
[the remark this answers is not in the document]
Comment (Kinplemelon) [reply src/views/Base/Login.vue:9]: 用来进行神奇的布局调整 `adjustLayout`

File: src/views/Base/Login.vue [hunk 8]
```diff
@@ -55,110 +55,108 @@
...
+
+  if (auto && username && token) redirect();
+  else toLogin(), setLanguage();
+
+  if (!auto) {
+    isLogining.value = true;
+    let res = await loginApi({
+      username,
+      password,
+    }).catch();
```
[the remark this answers is not in the document]
Comment (Kinplemelon) [reply src/views/Base/Login.vue:121]: 👌🏻

File: src/views/Dashboard/Metrics.vue [hunk 9]
```diff
@@ -139,237 +145,255 @@
...
+const updateStatsd = async function () {
+  statsdLoading.value = true;
+  let pendingData: MyStatsD = Object.assign({}, integrationData.statsd);
 
-    onMounted(() => {
-      metricsData();
-      loadIntegration();
+  Object.keys(pendingData).forEach((v) => {
+    if (v === "flush_time_interval" && pendingData[v] instanceof Array) {
+      pendingData[v] = (pendingData[v] as Array<string>).join("");
```
[the remark this answers is not in the document]
Comment (Kinplemelon) [reply src/views/Dashboard/Metrics.vue:361]: is [string, string]

[the remark this answers is not in the document]
Comment (kerry-emqx) [reply src/views/Dashboard/Metrics.vue:361]: then it is a bug elsewhere

**Review by kerry-emqx — APPROVED** [review]

**Pull request merged**